README
NAME
    SVN::Log::Index - Index and search over Subversion commit logs.

SYNOPSIS
      my $index = SVN::Log::Index->new({ index_path => '/path/to/index' });

      if($creating) {    # Create from scratch if necessary
        $index->create({ repo_url => 'url://for/repo' });
      }

      $index->open();    # And then open it

      # Now add revisions from the repo to the index
      $index->add({ start_rev => $start_rev,
                    end_rev   => $end_rev);

      # And query the index
      my $results = $index->search('query');

DESCRIPTION
    SVN::Log::Index builds a Plucene index of commit logs from a Subversion
    repository and allows you to do arbitrary full text searches over it.

METHODS
  new
      # Creating a new index object
      my $index = SVN::Log::Index->new({index_path => '/path/to/index'});

    Create a new index object.

    The single argument is a hash ref. Currently only one key is valid.

    index_path
        The path that contains (or will contain) the index files.

    This method prepares the object for use, but does not make any changes
    on disk.

  create
      $index->create({ repo_url       => 'url://for/repo',
                       analyzer_class => 'Plucene::Analysis::Analyzer::Sub',
                       optimize_every => $num,
                       overwrite      => 1, # or 0
                   });

    This method creates a new index, in the "index_path" given when the
    object was created.

    The single argument is a hash ref, with the following possible keys.

    repo_url
        The URL for the Subversion repository that is going to be indexed.

    analyzer_class
        A string giving the name of the class that will analyse log message
        text and tokenise it. This should derive from the
        Plucene::Analysis::Analyzer class. SVN::Log::Index will call this
        class' "new()" method.

        Once an analyzer class has been chosen for an index it can not be
        changed without deleting the index and creating it afresh.

        The default value is "Plucene::Analysis::SimpleAnalyzer".

    optimize_every
        Per the documentation for Plucene::Index::Writer, the index should
        be optimized to improve search performance.

        This is normally done after an application has finished adding
        documents to the index. However, if your application will be using
        the index while it's being updated you may wish the optimisation to
        be carried out periodically while the repository is still being
        indexed.

        If defined, the index will be optimized after every "optimize_every"
        revisions have been added to the index. The index is also optimized
        after the final revision has been added.

        So if "optimize_every" is given as 100, and you have requested that
        revisions 134 through 568 be indexed then the index will be
        optimized after adding revision 200, 300, 400, 500, and 568.

        The default value is 0, indicating that optimization should only be
        carried out after the final revision has been added.

    overwrite
        A boolean indicating whether or not a pre-existing index_path should
        be overwritten.

        Given this sequence;

          my $index = SVN::Log::Index->new({index_path => '/path'});
          $index->create({repo_url => 'url://for/repo'});

        The call to "create()" will fail if "/path" already exists.

        If "overwrite" is set to a true value then "/path" will be cleared.

    After creation the index directory will exist on disk, and a
    configuration file containing the create()-time parameters will be
    created in the index directory.

    Newly created indexes must still be opened.

  open
      $index->open();

    Opens the index, in preparation for adding or removing entries.

  add
      $index->add ({ start_rev      => $start_rev,  # number, or 'HEAD'
                     end_rev        => $end_rev,    # number, or 'HEAD'
                     optimize_every => $num });

    Add one or more log messages to the index.

    The single argument is a hash ref, with the following possible keys.

    start_rev
        The first revision to add to the index. May be given as "HEAD" to
        mean the repository's most recent (youngest) revision.

        This key is mandatory.

    end_rev
        The last revision to add to the index. May be given as "HEAD" to
        mean the repository's most recent (youngest) revision.

        This key is optional. If not included then only the revision
        specified by "start_rev" will be indexed.

    optimize_every
        Overrides the "optimize_every" value that was given in the
        "create()" call that created this index.

        This key is optional. If it is not included then the value used in
        the "create()" call is used. If it is included, and the value is
        "undef" then optimization will be disabled while these revisions are
        included.

        The index will still be optimized after the revisions have been
        added.

    Revisions from "start_rev" to "end_rev" are added inclusive. "start_rev"
    and "end_rev" may be given in ascending or descending order. Either:

      $index->add({ start_rev => 1, end_rev => 10 });

    or

      $index->add({ start_rev => 10, end_rev => 1 });

    In both cases, revisons are indexed in ascending order, so revision 1,
    followed by revision 2, and so on, up to revision 10.

  get_last_indexed_rev
      my $rev = $index->get_last_indexed_rev();

    Returns the revision number that was most recently added to the index.

    Most useful in repeated calls to "add()".

      # Loop forever.  Every five minutes wake up, and add all newly
      # committed revisions to the index.
      while(1) {
        sleep 300;
        $index->add({ start_rev => $index->get_last_indexed_rev() + 1,
                      end_rev   => 'HEAD' });
      }

    The last indexed revision number is saved as a property of the index.

  search
      my $hits = $index->search ($query);

    Search for $query (which is parsed into a Plucene::Search::Query object
    by the Plucene::QueryParser module) in $index and return a reference to
    an array of hash references. Each hash reference points to a hash where
    the key is the field name and the value is the field value for this hit.

    The keys are:

    relevance
        How relevant Plucene thought this result was, as a floating point
        number.

    url The URL of the repository that the index is for.

    revision, message, author, paths, date
        The revision number, log message, commit author, paths changed in
        the commit, and date of the commit, respectively.

QUERY SYNTAX
    This module supports the Lucene query syntax, described in detail at
    <http://lucene.apache.org/java/docs/queryparsersyntax.html>. A brief
    overview follows.

    *   A query consists of one or more terms, joined with boolean
        operators.

    *   A term is either a single word, or two or more words, enclosed in
        double quotes. So

          foo bar baz

        is a different query from

          "foo bar" baz

        The first searches for any of "foo", "bar", or "baz", the second
        searches for any of "foo bar", or "baz".

    *   By default, multiple terms in a query are OR'd together. You may
        also use "AND", or "NOT" between terms.

          foo AND bar
          foo NOT bar

        Use "+" before a term to indicate that it must appear, and "-"
        before a term to indicate that it must not appear.

          foo +bar
          -foo bar

    *   Use parantheses to control the ordering.

          (foo OR bar) AND baz

    *   Searches are conducted in *fields*. The default field to search is
        the log message. Other fields are indicated by placing the field
        name before the term, separating them both with a ":".

        Available fields are:

        revision
        author
        date
        paths

        For example, to find all commit messages where "nik" was the
        committer, that contained the string "foo bar":

          author:nik AND "foo bar"

SEE ALSO
    SVN::Log

BUGS
    Please report any bugs or feature requests to
    "bug-svn-log-index@rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SVN-Log-Index>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

AUTHOR
    The current maintainer is Nik Clayton, <nikc@cpan.org>.

    The original author was Garrett Rooney, <rooneg@electricjellyfish.net>

COPYRIGHT AND LICENSE
    Copyright 2006 Nik Clayton. All Rights Reserved.

    Copyright 2004 Garrett Rooney. All Rights Reserved.

    This software is licensed under the same terms as Perl itself.



CHANGES
[ Changes for 0.50 - Jan 13, 2006 ]

* Switch to KinoSearch instead of Plucene for indexing.  When indexing
  large repositories with thousands of repositories Plucene ran in to
  memory problems.  KinoSearch appears not to have this limitation.

[ Changes for 0.41 - Jan 23, 2006 ]

* Work around a Plucene bug by stringifying the revision number before
  indexing it.
* Fix a bug, include the repository URL in the results returned by search().
* Better documentation about the results returned by search().

[ Changes for 0.40 - Jan 20, 2006 ]

-- Changes by Nik Clayton

* Significant changes to the API.  It's not backwards compatible, but is
  now cleaner.
* One index now contains information about one repo.  It's now not possible
  to have one index contain information from multiple repos.
* Significant enhancements to the documentation.
* Require SVN::Log 0.03.  This gives us support for 'HEAD' in a revision
  specification.
* Indexing 'remembers' the last indexed revision.  The API exposes this,
  and bin/sli now supports 'LAST' in a revision specification.
* Significant changes to the bin/sli usage model.  It's now cleaner, but
  is not backwards compatible.

-- Changes by Garrett Rooney

* Accept an alternate Analyzer in constructor [Mike Henson <mike at axl.net>].
* Return the relevance in the result hash [Mike Henson <mike at axl.net>].

[ Changes for 0.30 - Jun 14, 2004 ]

* Changed verbose output of sli program to look more like 'svn log' output.
* Moved log retrieval code into SVN::Log module, distributed separately, 
  because it turns out that people want to use it.  In the process fixed a 
  bug in the command line client parsing code that caused us to miss revs.
* Add a mechanism for forcing use of the svn command line client, even if 
  the subversion perl bindings are installed.
* Add tests that exercise the code that uses the svn command line client 
  to read logs from the repository.

[ Changes for 0.22 - May 29, 2004 ]

* Fix the 'sli' script so that it can actually create a new index.

[ Changes for 0.21 - May 27, 2004 ]

* Fix the log parsing code, it was refering to a nonexistant variable.
* Make tests chmod the repository's format file since recent versions of 
  svnadmin make it read-only.

[ Changes for 0.20 - Mar 23, 2004 ]

* Added CHANGES file.
* Minor documentation fixes.
* Correct capitalization of Plucene in dependencies.
* If we don't have the SVN perl bindings installed, use the command line 
  client to get log entries from the repository instead.
* Explicitly look for the svn and svnadmin executables in Build.PL, instead 
  of just failing horribly if they aren't found.

[ Changes for 0.1 - Feb 9, 2004 ]

* Initial release.


lib/SVN/Log/Index.pm
package SVN::Log::Index;

use strict;
use warnings;

use File::Path;

use KinoSearch::InvIndexer;
use KinoSearch::Searcher;

use Params::Validate qw(:all);
use Exception::Class(
    'SVN::Log::Index::X::Args'
	=> { alias => 'throw_args', },
    'SVN::Log::Index::X::Fault'
	=> { alias  => 'throw_fault', },
);

Params::Validate::validation_options(
    on_fail => sub { throw_args error => shift },
);

use SVN::Log;
use YAML ();

our $VERSION = '0.5';

=head1 NAME

SVN::Log::Index - Index and search over Subversion commit logs.

=head1 SYNOPSIS

  my $index = SVN::Log::Index->new({ index_path => '/path/to/index' });

  if($creating) {    # Create from scratch if necessary
    $index->create({ repo_url => 'url://for/repo' });
  }

  $index->open();    # And then open it

  # Now add revisions from the repo to the index
  $index->add({ start_rev => $start_rev,
                end_rev   => $end_rev);

  # And query the index
  my $results = $index->search('query');

=head1 DESCRIPTION

SVN::Log::Index builds a KinoSearch index of commit logs from a
Subversion repository and allows you to do arbitrary full text searches
over it.

=head1 METHODS

=head2 new

  my $index = SVN::Log::Index->new({
      index_path => '/path/to/index'
  });

Create a new index object.

The single argument is a hash ref.  Currently only one key is valid.

=over 4

=item index_path

The path that contains (or will contain) the index files.

=back

This method prepares the object for use, but does not make any changes
on disk.

=cut

sub new {
  my $proto = shift;
  my $args  = validate(@_, {
      index_path => 1,
  });

  my $class = ref($proto) || $proto;
  my $self  = {};

  $self->{index_path} = $args->{index_path};

  bless $self, $class;
}

=head2 create

  $index->create({
      repo_url       => 'url://for/repo',
      analyzer_class => 'KinoSearch::Analysis::PolyAnalyzer',
      analyzer_opts  => [ language => 'en' ],
      overwrite      => 0, # Optional
  });

This method creates a new index, in the C<index_path> given when the
object was created.

The single argument is a hash ref, with the following possible keys.

=over 4

=item repo_url

The URL for the Subversion repository that is going to be indexed.

=item analyzer_class

A string giving the name of the class that will analyse log message
text and tokenise it.  This should derive from the
L<KinoSearch::Analysis::Analyzer> class.  SVN::Log::Index will call this
class' C<new()> method.

Once an analyzer class has been chosen for an index it can not be
changed without deleting the index and creating it afresh.

The default value is C<KinoSearch::Analysis::PolyAnalyzer>.

=item analyzer_opts

A list of options to be passed, as is, to the constructor for the
C<analyzer_class> object.

=item overwrite

A boolean indicating whether or not a pre-existing index_path should
be overwritten.

Given this sequence;

  my $index = SVN::Log::Index->new({index_path => '/path'});
  $index->create({repo_url => 'url://for/repo'});

The call to C<create()> will fail if C</path> already exists.

If C<overwrite> is set to a true value then C</path> will be cleared.

The default is false.

=back

After creation the index directory will exist on disk, and a
configuration file containing the create()-time parameters will be
created in the index directory.

Newly created indexes must still be opened.

=cut

sub create {
  my $self = shift;
  my $args = validate(@_, {
      repo_url => {
	  type => SCALAR,
	  regex => qr{^[a-z/]},
      },
      analyzer_class => {
	  type => SCALAR,
	  default => 'KinoSearch::Analysis::PolyAnalyzer',
      },
      analyzer_opts => {
	  type => ARRAYREF,
	  default => [ language => 'en' ],
      },
      overwrite => {
	  type => BOOLEAN,
	  default => 0,
      },
  });

  throw_fault("Can't call create() after open()")
      if exists $self->{config};

  if(-d $self->{index_path} and ! $args->{overwrite}) {
    throw_fault("create() $self->{index_path} exists and 'overwrite' is false");
  }

  if($args->{repo_url} !~ m/^(http|https|svn|file|svn\+ssh):\/\//) {
    $args->{repo_url} = 'file://' . $args->{repo_url};
  }

  $self->{config} = $args;
  $self->{config}{last_indexed_rev} = 0;

  $self->_create_analyzer();
  $self->_create_writer($args->{overwrite});

  $self->_save_config();

  delete $self->{config};	# Gets reloaded in open()
}

sub _save_config {
  my $self = shift;

  YAML::DumpFile($self->{index_path} . '/config.yaml', $self->{config})
      or throw_fault("Saving config failed: $!");
}

sub _load_config {
  my $self = shift;

  $self->{config} = YAML::LoadFile($self->{index_path} . '/config.yaml')
    or throw_fault("Could not load state from $self->{index_path}/config.yaml: $!");
}

sub _create_writer {
  my $self = shift;
  my $create = shift;

  return if exists $self->{writer} and defined $self->{analyzer};

  throw_fault("_create_analyzer() must be called first")
      if ! exists $self->{analyzer};
  throw_fault("analyzer is empty") if ! defined $self->{analyzer};

  $self->{writer} = KinoSearch::InvIndexer->new(
      invindex => $self->{index_path},
      create   => $create,
      analyzer => $self->{analyzer},
  ) or throw_fault("error creating writer: $!");

  foreach my $field (qw(paths revision author date message)) {
      $self->{writer}->spec_field(name => $field);
  }

  return;
}

sub _delete_writer {
    my $self = shift;
    my $optimize = shift;

    $self->{writer}->finish(optimize => $optimize);
    delete $self->{writer};
    return;
}

sub _create_analyzer {
  my $self = shift;

  return if exists $self->{analyzer} and defined $self->{analyzer};

  eval "require $self->{config}{analyzer_class}"
      or throw_fault "require($self->{config}{analyzer_class} failed: $!";

  $self->{analyzer} = $self->{config}{analyzer_class}->new(
      @{ $self->{config}{analyzer_opts} }
  ) or throw_fault("error creating $self->{config}{analyzer_class} object: $!");
}

=head2 open

  $index->open();

Opens the index, in preparation for adding or removing entries.

=cut

sub open {
  my $self = shift;
  my $args = shift;

  throw_fault("$self->{index_path} does not exist")
      if ! -d $self->{index_path};
  throw_fault("$self->{index_path}/config.yaml does not exist")
      if ! -f "$self->{index_path}/config.yaml";

  $self->_load_config();
  $self->_create_analyzer();
}

=head2 add

  $index->add ({
      start_rev      => $start_rev,  # number, or 'HEAD'
      end_rev        => $end_rev,    # number, or 'HEAD'
  });

Add one or more log messages to the index.

The single argument is a hash ref, with the following possible keys.

=over

=item start_rev

The first revision to add to the index.  May be given as C<HEAD> to mean
the repository's most recent (youngest) revision.

This key is mandatory.

=item end_rev

The last revision to add to the index.  May be given as C<HEAD> to mean
the repository's most recent (youngest) revision.

This key is optional.  If not included then only the revision specified
by C<start_rev> will be indexed.

=back

Revisions from C<start_rev> to C<end_rev> are added inclusive.
C<start_rev> and C<end_rev> may be given in ascending or descending order.
Either:

  $index->add({ start_rev => 1, end_rev => 10 });

or

  $index->add({ start_rev => 10, end_rev => 1 });

In both cases, revisons are indexed in ascending order, so revision 1,
followed by revision 2, and so on, up to revision 10.

=cut

sub add {
  my $self = shift;
  my $args = validate(@_, {
      start_rev => {
	  type => SCALAR
      },
      end_rev => {
	  type => SCALAR,
	  optional => 1
      },
  });

  $args->{end_rev} = $args->{start_rev} unless defined $args->{end_rev};

  foreach (qw(start_rev end_rev)) {
    throw_args("$_ value '$args->{$_}' is invalid")
      if $args->{$_} !~ /^(?:\d+|HEAD)$/;
  }

  # Get start_rev and end_rev in to ascending order.
  if($args->{start_rev} ne $args->{end_rev} and $args->{end_rev} ne 'HEAD') {
    if(($args->{start_rev} eq 'HEAD') or ($args->{start_rev} > $args->{end_rev})) {
      ($args->{start_rev}, $args->{end_rev}) =
	($args->{end_rev}, $args->{start_rev});
    }
  }

  $self->_create_writer(0);

  SVN::Log::retrieve ({ repository => $self->{config}{repo_url},
                        start      => $args->{start_rev},
                        end        => $args->{end_rev},
                        callback   => sub { $self->_handle_log({ rev => \@_ }) }
		    });

  $self->_delete_writer(1);

  return 1;
}

sub _handle_log {
  my ($self, $args) = @_;

  my ($paths, $rev, $author, $date, $msg) = @{$args->{rev}};

  my $doc = $self->{writer}->new_doc();

  $doc->set_value(revision => $rev);

  # it's certainly possible to get a undefined author, you just need either
  # mod_dav_svn with no auth, or svnserve with anonymous write access turned
  # on.
  $doc->set_value(author => $author) if defined $author;

  # XXX might want to convert the date to something more easily searchable,
  # but for now let's settle for just not tokenizing it.
  $doc->set_value(date => $date);

  $doc->set_value(paths => join(' ', keys %$paths))
    if defined $paths; # i'm still not entirely clear how this can happen...

  $doc->set_value(message => $msg)
    unless $msg =~ m/^\s*$/;

  $self->{writer}->add_doc($doc);

  $self->{config}{last_indexed_rev} = $rev;

  $self->_save_config();

  return;
}

=head2 get_last_indexed_rev

  my $rev = $index->get_last_indexed_rev();

Returns the revision number that was most recently added to the index.

Most useful in repeated calls to C<add()>.

  # Loop forever.  Every five minutes wake up, and add all newly
  # committed revisions to the index.
  while(1) {
    sleep 300;
    $index->add({ start_rev => $index->get_last_indexed_rev() + 1,
                  end_rev   => 'HEAD' });
  }

The last indexed revision number is saved as a property of the index.

=cut

sub get_last_indexed_rev {
  my $self = shift;

  throw_fault("Can't call get_last_indexed_rev() before open()")
    unless exists $self->{config};
  throw_fault("Empty configuration") unless defined $self->{config};

  return $self->{config}{last_indexed_rev};
}

=head2 search

  my $hits = $index->search($query);

Search for $query and returns a KinoSearch::Search::Hits object which
contains the result.

=cut

sub search {
  my ($self, $query) = @_;

  throw_fault("open() must be called first")
      unless exists $self->{config};

  my $searcher = KinoSearch::Searcher->new(
      invindex => $self->{index_path},
      analyzer => $self->{analyzer},
  );

  return $searcher->search(query => $query);
}

=head1 QUERY SYNTAX

This module supports the Lucene query syntax, described in detail at
L<http://lucene.apache.org/java/docs/queryparsersyntax.html>.  A brief
overview follows.

=over

=item *

A query consists of one or more terms, joined with boolean operators.

=item *

A term is either a single word, or two or more words, enclosed in double
quotes.  So

  foo bar baz

is a different query from

  "foo bar" baz

The first searches for any of C<foo>, C<bar>, or C<baz>, the second
searches for any of C<foo bar>, or C<baz>.

=item *

By default, multiple terms in a query are OR'd together.  You may also
use C<AND>, or C<NOT> between terms.

  foo AND bar
  foo NOT bar

Use C<+> before a term to indicate that it must appear, and C<->
before a term to indicate that it must not appear.

  foo +bar
  -foo bar

=item *

Use parantheses to control the ordering.

  (foo OR bar) AND baz

=item *

Searches are conducted in I<fields>.  The default field to search is
the log message.  Other fields are indicated by placing the field name
before the term, separating them both with a C<:>.

Available fields are:

=over

=item revision

=item author

=item date

=item paths

=back

For example, to find all commit messages where C<nik> was the committer,
that contained the string "foo bar":

  author:nik AND "foo bar"

=back

=head1 DIAGNOSTICS

Any of these methods may fail.  If they do, they throw an
L<Exception::Class> subclass representing the error, trappable with
C<eval>.  Uncaught exceptions will cause the client application to
C<die>.

=head2 SVN::Log::Index::X::Args

Represents an error that occurs if the parameters given to any of the
methods are wrong.  This might be because there are too few or too many
parameters, or that the types of those parameters are wrong.

The text of the error can be retrieved with the C<error()> method.

=head2 SVN::Log::Index::X::Fault

Represents any other error.

=head2 Example

  my $e;
  eval { $index->search('query string'); };

  if($e = SVN::Log::Index::X::Fault->caught()) {
      print "An error occured: ", $e->string(), "\n";
  } elsif ($e = Exception::Class->caught()) {
      # Something else failed, rethrow the error
      ref $e ? $e->rethrow() : die $e;
  }

=head1 SEE ALSO

L<SVN::Log>, L<KinoSearch>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-svn-log-index@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SVN-Log-Index>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 AUTHOR

The current maintainer is Nik Clayton, <nikc@cpan.org>.

The original author was Garrett Rooney, <rooneg@electricjellyfish.net>

=head1 COPYRIGHT AND LICENSE

Copyright 2006-2007 Nik Clayton.  All Rights Reserved.

Copyright 2004 Garrett Rooney.  All Rights Reserved.

This software is licensed under the same terms as Perl itself.

=cut

1;


Build.PL
#!/usr/bin/perl -w

# $Id: /local/CPAN/SVN-Log-Index/trunk/Build.PL 1474 2007-01-13T21:14:25.326886Z nik  $

use strict;
use warnings;

use Module::Build;

use Config;
use File::Spec;

sub have_exe {
  my $name = shift;

  my @dirs = split ':', $ENV{PATH};

  my $found_exe = 0;

  for (@dirs) {
    $found_exe = 1 if (-x File::Spec->catdir ($_, "$name$Config{_exe}"));
  }

  return $found_exe;
}

unless (have_exe ('svnadmin')) {
  print <<EOM;
SVN::Log::Index requires the 'svnadmin' executable to run it's tests.

You don't seem to have it, so don't be too surprised when they fail.
EOM
}

Module::Build->new (
  dist_name => 'SVN-Log-Index',

  license => 'perl',

  module_name => 'SVN::Log::Index',

  create_makefile_pl => 'passthrough',

  create_readme => 1,

  sign => 1,

  script_files => [
    'bin/sli',
  ],

  requires => {
    'KinoSearch' => 0.15,
    'SVN::Log' => 0.03,
    'Exception::Class' => 0,
    'YAML' => 0,
    'Test::Exception' => 0,
  },

  recommends => {
    'SVN::Core' => 0,
  },
)->create_build_script;
